database/pgx/pgx.go

```diff
@@ -211,11 +222,20 @@ func (p *Postgres) Open(url string) (database.Driver, error) {
 }
 
 func (p *Postgres) Close() error {
-	connErr := p.conn.Close()
-	dbErr := p.db.Close()
-	if connErr != nil || dbErr != nil {
-		return fmt.Errorf("conn: %v, db: %v", connErr, dbErr)
+	if p.db != nil {
+		err := stdlib.ReleaseConn(p.db, p.conn)
+		if err != nil {
+			return fmt.Errorf("release: %+v", err)
+		}
+
+		return nil
 	}
+
+	dbErr := p.conn.Close(context.Background())
```

Comment:
There are 3 cases `Close()` needs to handle:
1. Created with `Open()`
    * `p.db` and `p.conn` are both non-`nil`
    * should release the connection and close the DB
2. Created with `WithInstance()`
    * `p.db` and `p.conn` are both non-`nil`
    * _Should_ release the connection but leave the DB open for the caller of `WithInstance()` to close. However, this isn't the current behavior which is to close `p.db`, and existing clients may expect the current behavior, so let's leave it as is. e.g. close both `p.db` and release `p.conn`.
3. Created with `WithConnection()`
    * `p.db` is `nil` and `p.conn` is non-`nil`
    * Should leave the connection for the caller of `WithConnection()` to release the connection.

Per this [comment](https://github.com/golang-migrate/migrate/pull/593#discussion_r662378216), I think we should be releasing connection and never closing it